rego/rego.go

```diff
@@ -545,24 +564,127 @@ func (r *Rego) Compile(ctx context.Context) (*CompileResult, error) {
 	return result, nil
 }
 
-func (r *Rego) parse() (map[string]*ast.Module, ast.Body, error) {
+// Prepare will parse inputs, modules, and query arguments in preparation
+// of evaluating them.
+func (r *Rego) Prepare(ctx context.Context) (*PartialResult, error) {
```

Comment:
I was expecting Prepare to take options (e.g., `rego.Partial()` to enable the partial evaluation optimization) that could customize the Prepare step. We don't have to tackle this immediately, but worth considering (since I could imagine refactoring the `rego.Rego#PartialResult` function to be part of the prepare implementation.

This doesn't have to block the merge.

Reply:
We can do that. I initially backed off from adding parameters to keep it more simple and focused on the Eval use-case. I do agree we want it to be flexible enough to support the other API's so it makes sense to add it in now rather than break the API later.

Reply:
I took a swing at this in the latest update. I'm not totally happy with the way it turned out, but I think its an improvement.